encoding/iterators.go

```diff
@@ -0,0 +1,182 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package encoding
+
+import (
+	"time"
+
+	"github.com/m3db/m3db/ts"
+	"github.com/m3db/m3x/time"
+)
+
+var (
+	timeMax = time.Unix(1<<63-62135596801, 999999999)
+)
+
+// iterators is a collection of iterators, and allows for reading in order values
+// from the underlying iterators that are separately in order themselves.
+type iterators struct {
+	values      []Iterator
+	earliest    Iterator
+	earliestIdx int
+	earliestAt  time.Time
+	filtering   bool
+	filterStart time.Time
+	filterEnd   time.Time
+}
+
+func (i *iterators) len() int {
+	return len(i.values)
+}
+
+func (i *iterators) current() (ts.Datapoint, xtime.Unit, ts.Annotation) {
+	return i.earliest.Current()
+}
+
+func (i *iterators) at() time.Time {
+	return i.earliestAt
+}
+
+func (i *iterators) push(iter Iterator) bool {
+	if i.filtering && !i.moveIteratorToFilterNext(iter) {
+		return false
+	}
+	i.values = append(i.values, iter)
+	dp, _, _ := iter.Current()
+	if dp.Timestamp.Before(i.earliestAt) {
+		i.earliest = iter
+		i.earliestIdx = len(i.values) - 1
+		i.earliestAt = dp.Timestamp
+	}
+	return true
+}
+
+func (i *iterators) moveIteratorToFilterNext(iter Iterator) bool {
+	next := true
+	for next {
+		dp, _, _ := iter.Current()
+		if dp.Timestamp.Before(i.filterStart) {
+			// Filter out any before start
+			next = iter.Next()
+			continue
+		}
+		if !dp.Timestamp.Before(i.filterEnd) {
+			// Filter out completely if after end
+			next = false
+			break
+		}
+		// Within filter
+		break
+	}
+	return next
+}
+
+func (i *iterators) moveToValidNext() (bool, error) {
+	n := len(i.values)
+	if n == 0 {
+		return false, nil
+	}
+
+	var (
+		prevAt = i.earliestAt
+		next   = i.earliest.Next()
+	)
+	if i.filtering && next {
+		// Filter out values if applying filters
+		next = i.moveIteratorToFilterNext(i.earliest)
+	}
+
+	err := i.earliest.Err()
+	if err != nil {
+		i.reset()
+		return false, err
+	}
+
+	if n == 1 {
+		if next {
+			dp, _, _ := i.earliest.Current()
+			i.earliestAt = dp.Timestamp
+		} else {
+			i.reset()
+		}
+		return i.validateNext(next, prevAt)
+	}
+
+	if !next {
+		// No next so swap tail in and shrink by one
+		i.earliest.Close()
+		i.values[i.earliestIdx] = i.values[n-1]
+		i.values[n-1] = nil
+		i.values = i.values[:n-1]
+		n = n - 1
+	}
+
+	// Evaluate new earliest
+	i.earliest = i.values[0]
+	i.earliestIdx = 0
+	dp, _, _ := i.earliest.Current()
+	i.earliestAt = dp.Timestamp
+	for idx := 1; idx < n; idx++ {
+		dp, _, _ = i.values[idx].Current()
+		if dp.Timestamp.Before(i.earliestAt) {
+			i.earliest = i.values[idx]
+			i.earliestIdx = idx
+			i.earliestAt = dp.Timestamp
+		}
+	}
+
+	// Apply filter to new earliest if necessary
+	if i.filtering {
+		inFilter := i.earliestAt.Before(i.filterEnd) &&
```

Comment:
I wonder if it's more efficient to introduce a `filteredNext` method on the inner iterators to return the next value whose timestamp fall within the time range (or nil/false otherwise) and use it to move inner iterators to their respective next values, so if the earliest iterator has many datapoints before the start time, we don't need to recurse many levels deep to find the actual earliest iterator